File: sdk/typescript/src/ByteArray.ts
import { hexToUint8, uint8ToHex, isHexString, stringToAddress } from './utils/converter';

/**
 * Represents a fixed size byte array.
 */
export default class ByteArray {
	bytes: Uint8Array;
	/**
	 * Creates a byte array.
	 * @param {number} fixedSize Size of the array.
	 * @param {Uint8Array|string} arrayInput Byte array or hex string.
	 */
	constructor(fixedSize: number, arrayInput: Uint8Array | string) {
		let rawBytes = arrayInput;
		if ('string' === typeof rawBytes){
			try {
				if (isHexString(rawBytes)) {
          rawBytes = hexToUint8(rawBytes);
        } else {
          rawBytes = stringToAddress(rawBytes);
        }
			} catch (e) {
				throw Error('byte array is not a valid hex string');
			}
		}

		if (fixedSize !== rawBytes.length)
			throw RangeError(`bytes was size ${rawBytes.length} but must be ${fixedSize}`);

		/**
		 * Underlying bytes.
		 * @type Uint8Array
		 */
		this.bytes = new Uint8Array(rawBytes);
	}

	/**
	 * Returns string representation of this object.
	 * @returns {string} String representation of this object
	 */
	toString(): string {
		return uint8ToHex(this.bytes);
	}
}

File: sdk/csharp/src/ByteArray.cs
using System;
using SymbolSdk.Utils;

namespace SymbolSdk
{
    /**
     * Represents a fixed size byte array.
     */
    public abstract class ByteArray
    {
        public uint Size { get; }
        public byte[] bytes { get; }

        /**
	     * Creates a byte array.
	     * @param {byte} fixedSize Size of the array.
	     * @param {byte[]?} arrayInput Byte array.
	     */
        protected ByteArray(byte fixedSize, object? arrayInput)
        {
            bytes = new byte[fixedSize];
            if (arrayInput == null)
            {
            }
            else if (arrayInput is byte[] bytes1)
            {
                Size = fixedSize;
                if (fixedSize != bytes1.Length) throw new Exception($"bytes was size {bytes1.Length} but must be {fixedSize}");
                bytes = bytes1;
            }
            else if (arrayInput is string str)
            {
                if (str.Length == 39)
                {
                    bytes = Converter.StringToAddress(str);
                }
                else
                {
                    var rawBytes = Converter.HexToBytes(str);
                    if (fixedSize != rawBytes.Length) throw new Exception($"bytes was size {rawBytes.Length} but must be {fixedSize}");
                    bytes = rawBytes;   
                }
            }
        }

        /**
	     * Creates a byte array.
	     * @param {byte} fixedSize Size of the array.
	     * @param {string} arrayInput hex string.
	     */
        protected ByteArray(int fixedSize, string hexstring)
        {
            var rawBytes = Converter.HexToBytes(hexstring);
            if (fixedSize != rawBytes.Length) throw new Exception($"bytes was size {rawBytes.Length} but must be {fixedSize}");
            bytes = rawBytes;
        }

        /**
	     * Returns string representation of this object.
	     * @returns {string} String representation of this object
	     */
        public override string ToString()
        {
            return Converter.BytesToHex(bytes);
        }
    }
}
